test/integ/test_async.py

```diff
@@ -235,3 +235,13 @@ def test_close_connection_with_running_async_queries(conn_cnx):
             cur.execute_async("select count(*) from table(generator(timeLimit => 5))")
         assert not con._all_async_queries_finished()
     assert len(con._done_async_sfqids) < 2 and con.rest is None
+
+
+def test_close_connection_with_completed_async_queries(conn_cnx):
+    with conn_cnx() as con:
+        with con.cursor() as cur:
+            cur.execute_async("select 1")
+            cur.execute_async("select 2")
+        time.sleep(5)
```

Comment:
Instead of doing these hard sleeps, I think doing a while checks like this would be faster most of the time:
https://github.com/snowflakedb/snowflake-connector-python/blob/2d78cb0f397af24cc118c8942c83a852441be6cd/test/integ/test_async.py#L88
Unless you don't want the result caching to kick in, in that case we should maybe split up https://github.com/snowflakedb/snowflake-connector-python/blob/2d78cb0f397af24cc118c8942c83a852441be6cd/src/snowflake/connector/connection.py#L1374 into a non-caching implementation and a caching wrapper on it. Or maybe `mock.patch.dict` the async caching dict could also disable modifying the cache 🤣